### PR TITLE
Cleanup DaemonSet templateGeneration

### LINKED ISSUE #49336: Clean up deprecated DaemonSet templateGeneration field
We already announced deprecation of `.spec.templateGeneration` of extensions/v1beta1 DaemonSet in 1.7. 

Moving forward in 1.8, we'll need to:
* Add an `deprecated.daemonset.template.generation=<generation>` annotation (tentative name) to support conversion between extensions/v1beta1 DaemonSet and apps/v1beta2 DaemonSet. 
* Keep existing `templateGeneration` behavior for backward compatibility. 

@kubernetes/sig-apps-misc

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #49336 

**Special notes for your reviewer**:

/cc @janetkuo @foxish 

~~Depends on #49071~~ (Merged) 

**Release note**:

```release-note
None
```

### CODE REVIEW

File: pkg/apis/apps/v1beta2/conversion.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: No need to delete this annotation

File: pkg/apis/apps/v1beta2/conversion.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: They're pointers, need to be converted with the way you did in L393

File: pkg/apis/apps/v1beta2/conversion.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: Would you move these closer to other DaemonSet conversion codes? The TODO comment does not apply to these conversion codes.

**Comment by janetkuo** [comment]
Thanks a lot for the fix! You may need to run auto-gen scripts again. Please put those auto-generated codes in a separate commit in this PR so that rebase would be easier.

**Comment by janetkuo** [comment]
@k8s-bot ok to test

File: pkg/apis/apps/v1beta2/conversion.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: Need to check if `Annotations` is `nil` or not, otherwise this will panic

Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]:
```go
if out.Annotations == nil {
	out.Annotations = make(map[string]string)
}
```

File: pkg/apis/apps/v1beta2/conversion.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: Just took a closer look, you touched `out.Annotations` here, which is correct. Please ignore the previous comment.

Comment (dixudx) [inline pkg/apis/apps/v1beta2/conversion.go]: @janetkuo Will rebase and update the PR once #49071 gets merged. Seems #49071 has been approved.

**Comment by janetkuo** [comment]
@dixudx would you update conversion_test.go as well?

**Comment by dixudx** [comment]
@janetkuo Yeah, count on me.

**Comment by dixudx** [comment]
@janetkuo Add `conversion_test.go` in this PR or open a new PR ?

File: pkg/apis/apps/v1beta2/conversion.go
Comment (caesarxuchao) [inline pkg/apis/apps/v1beta2/conversion.go]: It looks like there is no generated conversions functions for the Status, could you check why?

You perhaps need to add a new comment `// +k8s:conversion-gen=k8s.io/kubernetes/pkg/apis/extensions` in `k8s.io/kubernetes/pkg/apis/v1beta2/doc.go`

Comment (caesarxuchao) [inline pkg/apis/apps/v1beta2/conversion.go]: Our roundtrip test doesn't exercise these conversions, because the internal Kind and the external Kind are of different group (see [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip/roundtrip.go#L210-L212)). We need to update `roundTripToAllExternalVersions` to test cross-group conversions. 

@sttts @deads2k @lavalamp do we have existing tests that exercise cross-group conversions?

**Comment by janetkuo** [comment]
> Add conversion_test.go in this PR or open a new PR ?

@dixudx I suggested updating conversion_test.go following the steps mentioned here https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#edit-conversiongo, but it seems that most apis don't have conversion_test.go. You can leave it for now.

File: pkg/apis/apps/v1beta2/conversion.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: Where is `// +k8s:conversion-gen` documented?

Comment (caesarxuchao) [inline pkg/apis/apps/v1beta2/conversion.go]: I didn't find any when i used it. We can doc it somewhere after @dixudx confirm what causes the missing generated conversions.

**Comment by janetkuo** [comment]
> Add conversion_test.go in this PR or open a new PR ?

@dixudx it can be done in a follow-up PR

**Comment by dixudx** [comment]
@janetkuo Already started in #49645.

File: pkg/apis/apps/v1beta2/conversion.go
Comment (dixudx) [inline pkg/apis/apps/v1beta2/conversion.go]: @caesarxuchao @janetkuo It works. After adding that, `Convert_v1beta2_DaemonSetStatus_To_extensions_DaemonSetStatus` and `Convert_extensions_DaemonSetStatus_To_v1beta2_DaemonSetStatus` are auto-generated. I would like to add this in another PR. WDYT?

**Comment by janetkuo** [comment]
@dixudx thanks, you're awesome!

File: pkg/apis/apps/v1beta2/conversion.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta2/conversion.go]: @dixudx I filed #49751 for both apps/v1beta1 and apps/v1beta2

**Comment by dixudx** [comment]
@janetkuo @caesarxuchao Rebased. PTAL. Thanks.

**Review by janetkuo — APPROVED** [review]

**Comment by dixudx** [comment]
@janetkuo @caesarxuchao Needs your `approval`. Thanks.

**Comment by janetkuo** [comment]
@dixudx I already approved this PR. Would you rebase again? I'll find someone to approve this PR and get it merged.

**Comment by dixudx** [comment]
@janetkuo Updated. PTAL.

**Comment by janetkuo** [comment]
Still lgtm.

File: docs/api-reference/apps/v1beta2/definitions.html
Comment (dixudx) [inline docs/api-reference/apps/v1beta2/definitions.html]: @janetkuo Seems weird here that this line is removed after running auto-generated.

Comment (janetkuo) [inline docs/api-reference/apps/v1beta2/definitions.html]: This is fine. Someone probably removed it from auto-gen code to avoid unnecessary rebases.

**Comment by bgrant0607** [comment]
I don't think a user will understand the change just by reading the release note.

https://github.com/kubernetes/community/issues/484

**Comment by dixudx** [comment]
@bgrant0607 Updated the release note. PTAL.

**Comment by janetkuo** [comment]
I don't recall what release note you wrote originally, but I think something like "Deprecate .spec.templateGeneration field of extensions/v1beta1 DaemonSet" is enough

**Comment by bgrant0607** [comment]
@janetkuo The original note was something like "Cleanup templateGeneration", which was vague. Your suggestion is fine.

**Comment by dixudx** [comment]
@janetkuo Rebased. PTAL. Thanks.

**Comment by janetkuo** [comment]
Just took a closer look, template generation was already deprecated in 1.7. This PR just removes the use of it in apps/v1beta2 DaemonSet. Since apps/v1beta2 was added in this release (1.8), this PR does not need a release note.

**Comment by dixudx** [comment]
/assign @thockin 
Need your `approval`. Thanks.

**Comment by janetkuo** [comment]
@bgrant0607 would you take another look?

**Comment by thockin** [comment]
Was v1beta2 available in 1.7 releases?

https://kubernetes.io/docs/reference/deprecation-policy/

**Comment by dixudx** [comment]
@thockin `apps/v1beta2` is for 1.8 release.

**Pull request merged**
